update event broker docs (#5378)
RasaHQ/rasa: docs/api/event-brokers.rst

.. edit-link::

> When running a Rasa Open Source assistant in production, you might want to

Reviewer: I'd suggest making this intro a bit clearer by making 'event broker' the subject of the first sentence. It feels a bit buried by the context. Something like this?

An event broker allows you to connect your assistant to other services that need access to data coming in from conversations. For example, you could connect your assistant to Rasa X to review and annotate conversations or forward messages to an external analytics service.

Author: Love it

connect your assistant to other services so that they can analyze the data

> that comes from your conversations. For example, you could `connect your

Reviewer 2: possibly to bypass the "data" singular/plural question

```suggestion
coming in from your conversations. For example, you could `connect your
```

live assistant to Rasa X <https://rasa.com/docs/rasa-x/installation-and-setup/existing-deployment/>`_
to review and annotate conversations, or forward messages to an external analytics
service. 

To do so, Rasa allows you to stream events to an event broker. The event broker
uses a message streaming service, also known as a message broker,

> to forward Rasa :ref:`events` from the Rasa server to other services.

Reviewer 2: I don't think we need to differentiate between "event broker" and "message broker", and probably pick one and use it throughout. they're synonymous IMO

Author: See this thread: https://rasa-hq.slack.com/archives/C7FCS16SD/p1583332500132700?thread_ts=1583284919.127600&cid=C7FCS16SD, people don't know what an event broker is -- i'll let @karen-white decide on the phrasing here

.. contents::
   :local:
   :depth: 1

Format
------

All events are streamed to the broker as serialised dictionaries every time

> the tracker updates it state. An example event emitted from the ``default``

Reviewer: go ahead an update `it state` to `its state`?

Author: good catch!

> The ``event`` field takes the event's ``type_name`` (for more on event

Author: I think this json format is old as well, does it need some metadata?

Reviewer 2: yes, `bot` events always have `data` and `metadata`

types, check out the :ref:`events` docs).

.. _event-brokers-pika:

Pika Event Broker
-----------------

The example implementation we're going to show you here uses `Pika <https://pika.readthedocs.io>`_,
the Python client library for `RabbitMQ <https://www.rabbitmq.com>`_.

.. contents::
   :local:

Adding a Pika Event Broker Using the Endpoint Configuration
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

You can instruct Rasa to stream all events to your Pika event broker by adding an ``event_broker`` section to your
``endpoints.yml``:

.. literalinclude:: ../../data/test_endpoints/event_brokers/pika_endpoint.yml

Rasa will automatically start streaming events when you restart the Rasa server.

> Adding a Pika Event Broker in Python

Author: These python additions look v out of date, why does this one pull from rasa platform code?

    pika_broker = PikaEventBroker('localhost',
                                  'username',
                                  'password',

>                                   queue='rasa_events')

Reviewer 2: the default queue is still `rasa_core_events`, but it makes sense to use this one in the docs for someone setting this up from scratch

Author: yeah, i was wondering how this even got set up. bc don't they have to configure the queue in their endpoints? Maybe it's not super clear that that needs to match this one if that is the case

Author: @ricwo what do you mean by `default queue` btw? if you don't assign a queue in your endpoints does it create one?

Reviewer 2: that's right, if you don't specify `queue` it uses this the default kwarg

>     $ pip install rasa[kafka]

Author: is still still necessary and valid?

Reviewer 2: nope, it's no longer optional (it's always installed since poetry was introduced)

.. contents::
   :local:

Adding a Kafka Event Broker Using the Endpoint Configuration
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> You can instruct Rasa to stream all events to your Kaf event broker by adding an ``event_broker`` section to your

Reviewer 2: ```suggestion
You can instruct Rasa to stream all events to your Kafka event broker by adding an ``event_broker`` section to your
```

Author: 🙈
